File: src/main/java/com/leijendary/spring/boot/template/specification/SampleListSpecification.java
package com.leijendary.spring.boot.template.specification;

import static com.leijendary.spring.boot.template.specification.TranslationPredicate.query;
import static com.leijendary.spring.boot.template.util.PredicateUtil.lowerLike;
import static org.apache.commons.lang3.StringUtils.isBlank;

import javax.persistence.criteria.CriteriaBuilder;
import javax.persistence.criteria.CriteriaQuery;
import javax.persistence.criteria.Predicate;
import javax.persistence.criteria.Root;

import com.leijendary.spring.boot.template.model.SampleTable;

import org.springframework.data.jpa.domain.Specification;
import org.springframework.lang.NonNull;

import lombok.Builder;

@Builder
public class SampleListSpecification implements Specification<SampleTable> {

    private final String query;

    @Override
    public Predicate toPredicate(@NonNull final Root<SampleTable> root, @NonNull final CriteriaQuery<?> criteriaQuery,
                                 @NonNull final CriteriaBuilder criteriaBuilder) {
        if (isBlank(query)) {
            return criteriaQuery.where().getRestriction();
        }

        // Column 1 filtering
        final var column1Path = root.<String>get("column1");
        final var column1Like = lowerLike(query, column1Path, criteriaBuilder);
        final var translations = query(root, criteriaBuilder, query, "name", "description");

        final var predicate = criteriaBuilder.or(column1Like, translations);

        return criteriaQuery.where(predicate).getRestriction();
    }
}

File: src/main/kotlin/com/leijendary/spring/template/specification/SampleListSpecification.kt
package com.leijendary.spring.template.specification

import com.leijendary.spring.template.core.extension.lowerLike
import com.leijendary.spring.template.entity.SampleTable
import jakarta.persistence.criteria.CriteriaBuilder
import jakarta.persistence.criteria.CriteriaQuery
import jakarta.persistence.criteria.Predicate
import jakarta.persistence.criteria.Root
import org.springframework.data.jpa.domain.Specification

class SampleListSpecification(private val query: String? = null) : Specification<SampleTable> {
    override fun toPredicate(
        root: Root<SampleTable>,
        criteriaQuery: CriteriaQuery<*>,
        criteriaBuilder: CriteriaBuilder
    ): Predicate {
        if (query.isNullOrBlank()) {
            return criteriaBuilder.and()
        }

        // Column 1 filtering
        val column1Path = root.get<String>("column1")
        val column1Like = lowerLike(query, column1Path, criteriaBuilder)

        return criteriaQuery.where(column1Like).restriction
    }
}
